### PR TITLE
[Feature] Mask Replay

### PR DESCRIPTION
**Pull request by vx120**
## Summary

This PR adds experimental support for **sampling distribution replay**.

A sampling mask represents the vocabulary support retained after top-k/top-p filtering. It is not an attention mask and does not affect causal attention or KV-cache behavior.

When enabled, vLLM returns the sampling support for each generated token in a CSR-style representation:

```python
SamplingMask(
    token_ids=[...],
    offsets=[0, ..., len(token_ids)],
)
```

The feature is opt-in and does not change default generation behavior.

## Motivation

With top-p sampling, rollout probabilities are normalized over a truncated vocabulary, while training commonly recomputes probabilities over the full vocabulary. This creates a systematic mismatch in importance ratios and KL estimates.

Sampling replay returns the rollout-time support so training can recompute log-probabilities over the same distribution.

## Experiment

We compare GRPO training with sampling replay enabled and disabled under top-p sampling.

For each generated token:

- `logp_diff = training_logprob - rollout_logprob`
- `importance_ratio = exp(logp_diff)`

Ideally, the rollout and training distributions match, so `logp_diff` is close to `0` and `importance_ratio` is close to `1`.

Sampling replay metrics

### Mean Importance Ratio

With replay enabled, the mean importance ratio remains tightly centered around `1`. Without replay, it is consistently below `1` and shows larger fluctuations.

This indicates that replay restores the rollout-time top-p normalization during training, while the non-replay baseline compares the truncated rollout distribution with a full-vocabulary training distribution.

### Mean Log-Probability Difference

With replay enabled, the mean log-probability difference remains close to `0`. Without replay, it has a persistent negative bias.

This is expected: a token probability normalized over the full vocabulary is generally lower than the probability normalized over the retained top-p support.

### Standard Deviation of the Log-Probability Difference

Replay lowers the standard deviation of the log-probability difference. This means it improves not only the average alignment, but also token-level consistency between rollout and training probabilities.

Lower variance produces more stable importance weights and reduces the impact of probability-ratio outliers.

### Approximate KL

Replay produces a lower and more stable approximate KL estimate, with fewer large spikes. This indicates that the training-time distribution stays closer to the rollout-time distribution throughout optimization.

### Conclusion

Across all four metrics, sampling replay improves rollout/training distribution consistency:

- `importance_ratio_mean` stays close to `1`;
- `logp_diff_mean` stays close to `0`;
- `logp_diff_std` is lower;
- `approx_kl` is lower and more stable.

These results demonstrate the intended algorithmic effect of sampling replay: improved probability alignment and lower-variance importance weights.

This experiment does not, by itself, establish an improvement in downstream reward or task accuracy.

## Limitations

The current implementation requires:

- Model Runner V2;
- `logprobs_mode="processed_logprobs"`;
- `SamplingParams.logprobs=1`;
- `temperature > 0`.

Speculative decoding, custom samplers, and custom logits processors are not currently supported.

### CODE REVIEW

**Comment by aoshen02** [comment]
@codex review

File: vllm/v1/engine/output_processor.py [hunk 1]
```diff
@@ -406,6 +409,15 @@ def _new_completion_output(
         if delta and logprobs:
             logprobs = logprobs[-len(token_ids) :]
 
+        sampling_mask = None
+        if (
+            finished
```
Comment (aoshen02) [inline vllm/v1/engine/output_processor.py]: Is it for reducing the transfer cost? Maybe we should let it be final only mode if this feature is enabled.

File: vllm/v1/core/sched/scheduler.py [hunk 2]
```diff
@@ -332,7 +332,6 @@ def __init__(
         self.enable_return_routed_experts = (
             vllm_config.model_config.enable_return_routed_experts
         )
-
         if self.enable_return_routed_experts:
```
Comment (aoshen02) [inline vllm/v1/core/sched/scheduler.py:330]: Please don't modify here.

File: vllm/config/vllm.py [hunk 3]
```diff
@@ -938,6 +938,25 @@ def _verify_kv_transfer_compat(self) -> None:
             "expandable_segments is automatically disabled)."
         )
 
+    def _verify_sampling_replay_config(self) -> None:
+        model_config = self.model_config
+        if model_config is None or not model_config.enable_return_sampling_mask:
+            return
+        if not self.use_v2_model_runner:
+            raise ValueError("sampling distribution replay requires Model Runner V2")
+        if self.speculative_config is not None:
```
Comment (aoshen02) [inline vllm/config/vllm.py:1048]: Will it be a follow up pr to support spec?

File: vllm/v1/core/sched/scheduler.py (hunk 2 above)
Comment (vx120) [reply vllm/v1/core/sched/scheduler.py:330]: ok

File: vllm/v1/engine/output_processor.py (hunk 1 above)
Comment (vx120) [reply vllm/v1/engine/output_processor.py]: I agree with you. I found the same question when I optimized the training code

File: vllm/config/vllm.py (hunk 3 above)
Comment (vx120) [reply vllm/config/vllm.py:1048]: We’ll evaluate the need after the current phase is complete

Comment (aoshen02) [reply vllm/config/vllm.py:1048]: Sure, I would strongly recommend doing this.

File: vllm/v1/core/sched/scheduler.py [hunk 4]
```diff
@@ -1907,6 +1908,12 @@ def update_from_output(
             ):
                 new_logprobs = logprobs.slice_request(req_index, len(new_token_ids))
 
+            sampling_masks = model_runner_output.sampling_masks
+            if new_token_ids and sampling_masks is not None:
+                new_sampling_mask = sampling_masks.slice_request(
+                    req_index, len(new_token_ids)
```
Comment (aoshen02) [inline vllm/v1/core/sched/scheduler.py]: Should use a flag to open it, or it would be in critical path.

File: vllm/v1/core/sched/scheduler.py [hunk 5]
```diff
@@ -1907,6 +1908,12 @@ def update_from_output(
             ):
                 new_logprobs = logprobs.slice_request(req_index, len(new_token_ids))
 
+            sampling_masks = model_runner_output.sampling_masks
+            if new_token_ids and sampling_masks is not None:
```
Comment (aoshen02) [inline vllm/v1/core/sched/scheduler.py]: if self.enable_return_sampling_mask and new_token_ids:

File: vllm/v1/worker/gpu/sample/sampling_mask.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+import torch
+
+from vllm.v1.worker.gpu.sample.output import SamplingMaskTensors
+
+
+def compact_sampling_mask(
```
Comment (aoshen02) [inline vllm/v1/worker/gpu/sample/sampling_mask.py]: I think build_sampling_keep_mask is more appropriate?

File: vllm/v1/worker/gpu/sample/sampling_mask.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    processed_logits: torch.Tensor,
```
Comment (aoshen02) [inline vllm/v1/worker/gpu/sample/sampling_mask.py]: We can inline this I think, as only one place call this.

File: vllm/v1/engine/input_processor.py [hunk 8]
```diff
@@ -327,6 +327,21 @@ def process_inputs(
...
                 sampling_params.update_from_tokenizer(self.tokenizer)
+            if self.model_config.enable_return_sampling_mask:
+                if sampling_params.output_kind != RequestOutputKind.FINAL_ONLY:
+                    raise ValueError(
+                        "sampling distribution replay requires output_kind=FINAL_ONLY"
+                    )
+                if sampling_params.temperature <= 0:
+                    raise ValueError(
+                        "sampling distribution replay requires temperature > 0"
+                    )
```
Comment (aoshen02) [inline vllm/v1/engine/input_processor.py]: Why don't put into _validate_params?

**Comment by aoshen02** [comment]
LGTM. @njhill could you please take a look?

File: vllm/v1/worker/gpu/sample/output.py [hunk 9]
```diff
@@ -1,10 +1,56 @@
...
+        cls,
+        logits: torch.Tensor,
+        num_sampled_tokens: torch.Tensor,
+    ) -> "SamplingMaskTensors":
+        """Compact the finite-logit support for requests that sampled tokens."""
+        keep = torch.isfinite(logits)[num_sampled_tokens.bool()]
+        counts = keep.sum(dim=-1, dtype=torch.int32)
+        token_ids = keep.flatten().nonzero(as_tuple=True)[0]
+        token_ids = token_ids.remainder(keep.shape[1]).to(torch.int32)
+        return cls(token_ids, counts)
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/sample/output.py]: This causes GPU-CPU sync. Maybe you can use Triton to avoid that!

Comment (aoshen02) [reply vllm/v1/worker/gpu/sample/output.py]: > This causes GPU-CPU sync. Maybe you can use Triton to avoid that!

Oh, thanks for the comment.

File: vllm/sampling_mask.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+from dataclasses import dataclass
+
+
+@dataclass
+class SamplingMask:
```
Comment (njhill) [inline vllm/sampling_mask.py]: It seems a bit odd t have this in a file by itself. Perhaps better in the sibling outputs.py file? (or another if more appropriate)

File: vllm/v1/worker/gpu/sample/output.py [hunk 11]
```diff
@@ -1,10 +1,122 @@
...
+import numpy as np
 import torch
 
-from vllm.v1.outputs import LogprobsTensors
+from vllm.triton_utils import tl, triton
+from vllm.v1.outputs import LogprobsTensors, SamplingMaskLists
+
+
+@triton.jit
+def _pack_sampling_mask_kernel(
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/output.py:26]: nit: suggest to move the changes to this file to the end since they are somewhat niche

File: vllm/entrypoints/llm.py [hunk 12]
```diff
@@ -201,6 +203,7 @@ def __init__(
         offload_params: set[str] | None = None,
         enforce_eager: bool = False,
         enable_return_routed_experts: bool = False,
+        enable_return_sampling_mask: bool = False,
```
Comment (njhill) [inline vllm/entrypoints/llm.py]: I know this is consistent with the exiting `enable_return_routed_experts` but I feel that the `enable_` is kind of redundant and it would be better to name it just `return_sampling_mask` everywhere.

Perhaps we can do this, and separately (not this PR ofc) also rename the routed experts one (of course keeping/deprecating the old version of that for a while).

**Review by njhill — COMMENTED** [review]
Thanks @vx120 the changes look pretty clean to me!

File: vllm/entrypoints/llm.py (hunk 12 above)
Comment (aoshen02) [reply vllm/entrypoints/llm.py]: done, will rename enable_return_routed_experts in our r3 pr.

File: vllm/sampling_mask.py (hunk 10 above)
Comment (aoshen02) [reply vllm/sampling_mask.py]: I see, I just don't feel like the samplingmask is the same type of class in the outputs.py file.

**Review by njhill — APPROVED** [review]
Thanks @vx120 @aoshen02
